python/lsst/pipe/tasks/script/registerDcrSubfilters.py

```diff
@@ -0,0 +1,60 @@
+# This file is part of pipe_tasks.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+import click
+from collections import defaultdict
+import logging
+from sqlalchemy.exc import IntegrityError
+
+from lsst.daf.butler import Butler
+
+
+_log = logging.getLogger(__name__.partition(".")[2])
+
+
+def registerDcrSubfilters(repo, num_subfilters, filter_names):
+    """Construct a set of subfilters for chromatic modeling and add them to a
+    registry.
+
+    Parameters
+    ----------
+    repo : `str`
+        URI to the location to read the repo.
+    num_subfilters : `int`
+        The number of subfilters to add.
+    filter_names : `list` [`str`]
+        The filter names to add.
+    """
+    butler = Butler(repo, writeable=True)
+    records = defaultdict(list)
+    with butler.registry.transaction():
+        record = []
+        for filterName in filter_names:
+            for sub in range(num_subfilters):
+                record.append({"band": filterName, "subfilter": sub})
+                records[filterName].append(sub)
+        try:
+            butler.registry.insertDimensionData("subfilter", *record)
+        except IntegrityError as err:
+            raise click.ClickException(f"Subfilters for at least one filter of {filter_names} "
+                                       f"are already defined in {repo}.") from err
```

Comment:
@isullivan, would it be more convenient if registering the same subfilter multiple times was a harmless no-op insteadof an error?  If so, we can just change this from `insertDimensionData` to `syncDimensionData` (and pass it one record at a time), and we'll get that behavior instead (with no need to `try/except`).

Reply:
That would be fine, as long as it prints a warning if the subfilter is already defined.

Reply:
@isullivan, @TallJimbo  I've changed from `insert` to `sync`. I had to make some changes to the user feedback to print the warning, please take a quick look at the code in this file. And here is an example of what running the command looks like right now, where there are already sub filters 0-2 for the bar filter, and it adds 3 and 4 to bar and 0-4 for baz. Does this seem good?

    $ butler register-dcr-subfilters foo 5 bar baz
    For filter "bar": registered subfilters [3, 4], subfilters [0, 1, 2] already existed.
    For filter "baz": registered subfilters [0, 1, 2, 3, 4].

Reply:
Thank you for the example. If "bar" already has subfilters [0, 1, 2] defined, then it should not register subfilters [3, 4] if someone runs `$ butler register-dcr-subfilters foo 5 bar baz`. The meaning of each subfilter depends on the number of subfilters, so adding [3, 4] in this case will change the meaning of the existing subfilters [0, 1, 2] and will break any data products in the repository that depend on them.
If a given band already has subfilters defined, it should be skipped entirely, and either raise an error like before or print a warning that it is being skipped. The output could look like:
```
$ butler register-dcr-subfilters foo 5 bar baz
For filter "bar": Not registering subfilters for "bar": subfilters [0, 1, 2] already existed.
For filter "baz": registered subfilters [0, 1, 2, 3, 4].
```

Reply:
Oh, that's my misunderstanding then, and it means both that we might not be able to use `syncDimensionData` here, and there's a potentially much bigger problem: once you've set the number of subfilters for a band in a repo (even a long-lived, shared repo), you can _never_ change it.

I was instead assuming that if you had more subfilters than you wanted for a band, you could ignore them in a particular processing run with e.g. `-d subfilter IN (0, 1)` (and then be super careful about comparing across collections, as the subfilter interpretation might differ).  And so adding new subfilters as needed is totally fine.

Reply:
Would there be a way to tie the number of subfilters to specific reruns? That would be best, since you would not want the number of subfilters to change within a rerun, but certainly might want to make different reruns with different numbers of subfilters.

Reply:
There's nothing special that would tie subfilters to reruns, but if there's one dataset that uses subfilters from which all others are are ultimately derived (e.g. a `dcrCoadd`), the pipeline that makes that dataset would need to be run with care to limit the subfilters to the ones desired, but downstream pipelines that start from those outputs would naturally pick them up (in the same sense that if you build coadds in just some patches, downstream processing on those would just include those patches).

Reply:
@TallJimbo, @isullivan It sounds like what I should do right now is go back to using `insertDimensionData`, and separate calls to it by band. That way if it fails for one band because of existing subfilters the function can continue registering the subfilters for other bands. And then it will report the outcome per Ian's example above. 
Is that right, or are you working out a different approach?

Reply:
@n8pease That sounds fine to me. In the future we should work out a way to un-register or over-write registered subfilters for a collection.

Reply:
Part of me wonders why we can't always register N subfilters where N is larger than anyone could ever need and then pick an M<=N for any processing run. Then you wouldn't need to ask people what N should be.

Reply:
@timj The problem is that the number subfilters defines their bandwidth. E.g. if there are 3 subfilters in g-band and the full band is 400-550nm, then that defines the wavelength ranges included in each subfilter as 400-450nm, 450-500nm, and 500-550nm.

Reply:
That's a shame. I was hoping you could say "For this processing I'm defining N to be 3" and have the processing split the band up appropriately. It seems that if we can't define it this way then once the number of subfilters are defined then changing them at any later time would break all previous processing.

Reply:
Let's let @n8pease get this merged and then come back to this question.  I don't think the situation we're in now is tenable, but I also suspect that there isn't a lot of urgency for making DCR coadds in big shared repos right now anyway, and hence we don't have to resolve it right away.